File: 0922_sort_array_by_parity_ii/sort_array_by_parity.java
class Solution {
    // Time: O(n)
    // Space: O(1)
    public int[] sortArrayByParityII(int[] nums) {
        int i = 0, j = 1, n = nums.length;
        while (i < n && j < n) {
            if (nums[i] % 2 == 0) {
                i += 2;
            } else if (nums[j] % 2 == 1) {
                j += 2;
            } else {
                int tmp = nums[i];
                nums[i] = nums[j];
                nums[j] = tmp;
                i += 2;
                j += 2;
            }
        }
        return nums;
    }
}

File: 0922_sort_array_by_parity_ii/sort_array_by_parity.py
from typing import List


def sortArrayByParityII(nums: List[int]) -> List[int]:
    """
    Time: O(n)
    Space: O(1)
    """
    i, j = 0, 1
    n = len(nums)
    while i < n and j < n:
        if nums[i] % 2 == 0:
            i += 2
        elif nums[j] % 2 == 1:
            j += 2
        else:
            nums[i], nums[j] = nums[j], nums[i]
            i += 2
            j += 2
    return nums
    
    